Samples/OpenQasmReader/Parser.cs

```diff
@@ -0,0 +1,942 @@
+        /// </summary>
+        /// <param name="token">Current token the tokenizer is on to parse</param>
+        /// <param name="cRegs">Classical registers defined</param>
+        /// <param name="path">Directory the qasm is located in (mostly for inlcude purposes)</param>
+        private static void ParseClassicalRegister(IEnumerator<string> token, Dictionary<string, int> cRegs, StringBuilder inside)
+        {
+            token.MoveNext();
+            var name = token.Current;
+            var index = name.IndexOf('[') + 1;
+            var count = int.Parse(name.Substring(index, name.IndexOf(']') - index));
+            name = name.Remove(index - 1);
+            cRegs.Add(name, count);
+            
+            token.MoveNext(); //;
+        }
+
+        private static void ParseQuantumRegister(IEnumerator<string> token, Dictionary<string, int> qRegs)
+        {
+            token.MoveNext();
+            var name = token.Current;
+            var index = name.IndexOf('[') +1;
+            var count = int.Parse(name.Substring(index, name.IndexOf(']') - index));
+            qRegs.Add(name.Remove(index-1), count);
+            token.MoveNext(); //;
+        }
+
+        private static void ParseGateCall(IEnumerator<string> token, StringBuilder builder, Dictionary<string,int> qReg)
+        {
+            var gateName = token.Current;
+            var doubles = new List<string>();
+            var qbits = new List<string>();
+            bool withinParentheses = false;
+            while (token.MoveNext() && !token.Current.Equals(POINT_COMMA))
+            {
+                if (token.Current.Equals(COMMA))
+                {
+                    continue;
+                }
+                else if (token.Current.Equals(CLOSE_PARANTHESES))
+                {
+                    withinParentheses = false;
+                }
+                else if (withinParentheses)
+                {
+                    doubles.Add(ParseCalculation(token, COMMA, CLOSE_PARANTHESES));
+                    if (token.Current.Equals(CLOSE_PARANTHESES))
+                    {
+                        withinParentheses = false;
+                    }
+                }
+                else if (token.Current.Equals(OPEN_PARANTHESES))
+                {
+                    withinParentheses = true;
+                }
+                else
+                {
+                    qbits.Add(token.Current);
+                }
+            }
+
+            var loopRequired = qReg.Count != 0 && qbits.Any() && !qbits.Any(q => q.Contains('['));
+            if (loopRequired)
+            {
+                builder.Append(INDENTED);
+                var size = qbits.First(q => !q.Contains('['));
+                builder.AppendFormat("for(_idx in 0..Length({0})){{\n", size);
+                builder.Append("    ");
+            }
+            builder.Append(INDENTED);
+            builder.Append(FirstLetterToUpperCase(gateName));
```

Comment:
is there a technical reason for this? i think it would be better to leave the user-defined gate names as-is, but more of a personal preference.

Reply:
got it.